File: Tools/FakeBuilder/FsQuery.fs
module Fake.FsQuery

open System
open System.IO
open System.Text.RegularExpressions

type PathPattern(pattern: string) = 
    inherit Regex(pattern.Replace("/","\\\\").Replace(".","\\.").Replace("*",".*"), RegexOptions.Compiled ||| RegexOptions.CultureInvariant ||| RegexOptions.ExplicitCapture ||| RegexOptions.IgnoreCase ||| RegexOptions.Singleline)


let toPatterns strings=
   strings |> Seq.toList |> List.map(fun x -> new PathPattern(x))
let tupleRelative baseDir (pathStrings:list<string>) =
    pathStrings |> 
    List.filter(fun item -> item.StartsWith(baseDir)) |>
    List.map(fun item -> (item, item.Substring(baseDir.Length).TrimEnd('\\', '/')))

type FsInventory(folder :string) =
  let mutable iFiles = []
  member this.baseDir = folder.TrimEnd('\\','/')
  member this.files
    with get () =
      if iFiles = [] then Directory.GetFiles(this.baseDir, "*", SearchOption.AllDirectories) |>  Array.toList |> tupleRelative this.baseDir
      else iFiles
    and set (value) = iFiles <- value
  member this.folders = Directory.GetDirectories(this.baseDir, "*", SearchOption.AllDirectories) |>  Array.toList |> tupleRelative(this.baseDir)

type FsQuery(inventory:FsInventory, exclusions: list<PathPattern>) = 
  let inv = inventory
  let excl = exclusions
  let query(baseDir:string, paths:list<string * string>, exclusions:list<PathPattern>, queries:list<PathPattern>)=
    paths |> List.filter 
      (fun (f,r) -> 
           (exclusions |> List.tryFind(fun e -> e.IsMatch(r)) = None) &&
           (queries |> List.tryFind(fun e -> e.IsMatch(r)) <> None)) 
          |> List.map (fun (f,r) -> f)

  
  member this.exclude([<ParamArray>] exclusions: string list) = 
    new FsQuery(inv,excl @ toPatterns exclusions)
  
  member this.exclude([<ParamArray>] exclusions: string[]) = 
    new FsQuery(inv,excl @ toPatterns exclusions)
  
  member this.exclude(exclusions: list<PathPattern>) = 
    new FsQuery(inv,excl @ exclusions)
  
  member this.files([<ParamArray>] queries: string list) =
    query(inv.baseDir,inv.files,excl,toPatterns queries)
  
  member this.files([<ParamArray>] queries: string[]) =
    query(inv.baseDir,inv.files,excl,toPatterns queries)
  
  member this.files([<ParamArray>] queries: PathPattern[]) =
    query(inv.baseDir,inv.files,excl, queries |> Array.toList)

  member this.folders([<ParamArray>] queries: string[]) =
    query(inv.baseDir,inv.folders,excl,toPatterns queries)
 
  member this.folders([<ParamArray>] queries: PathPattern[]) =
    query(inv.baseDir,inv.folders,excl, queries |> Array.toList)

File: Tools/BuildTools/FsQuery.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace BuildTools {
    /// <summary>
    /// Provides high-speed, cached, regex-capable filesystem query ability. 
    /// </summary>
    public class FsQuery {
        string[] _files;
        string[] _folders;
        string baseDir = null;

        public FsQuery(string dir):this(dir,null) {

        }

        public FsQuery(string dir, IEnumerable<string> exclude) {
            baseDir = dir.TrimEnd('\\','/');
            Rescan();
            exclusions = new List<Pattern>();
            if (exclude != null) AddExclusions(exclude);
           
        }

        public void Rescan()
        {
            _files = Directory.GetFiles(baseDir, "*", SearchOption.AllDirectories);
            _folders = Directory.GetDirectories(baseDir, "*", SearchOption.AllDirectories);
        }


        public void AddExclusions(IEnumerable<string> exclude) {
            foreach (string s in exclude) exclusions.Add(new Pattern(s));
        }

        public List<Pattern> exclusions = null;


        public List<string> files(params string[] p) {
            return query(_files, exclusions, toPatterns(p));
        }
        public List<string> folders(params string[] p) {
            return query(_folders, exclusions, toPatterns(p));
        }


        public List<string> files(params Pattern[] p){
            return query(_files, exclusions, p);
        }
        public List<string> folders(params Pattern[] p) {
            return query(_folders, exclusions, p);
        }

        public Pattern[] toPatterns(string[] patterns) {
            Pattern[] p = new Pattern[patterns.Length];
            for (int i = 0; i < patterns.Length; i++)
                p[i] = new Pattern(patterns[i]);
            return p;
        }

        protected List<String> query(string[] items, IEnumerable<Pattern> exclusions, IEnumerable<Pattern> queries) {
            Console.WriteLine("Applying queries:");
            foreach (Pattern p in queries) {
                Console.WriteLine(p.ToString());
            }
            List<string> results = new List<string>();
            foreach (string item in items) {
                //Trim to basedir
                if (!item.StartsWith(baseDir)) throw new Exception("Paths don't match: " + item + " , " + baseDir);
                string s = item.Substring(baseDir.Length).TrimEnd('\\', '/');

                bool excluded = false;
                if (exclusions != null)
                    foreach (Pattern e in exclusions)
                        if (e.IsMatch(s)) {
                            excluded = true;
                           // Console.WriteLine("Excluded " + s);
                            break;
                        }

                if (!excluded) {
                    foreach (Pattern q in queries) {
                        if (q.IsMatch(s)) {
                            results.Add(item);
                            break;
                        }
                    }
                }
            }
            return results;
        }


    }
}
